### DISCUSSION TITLE
Best way to display many objects on a Globus

### DISCUSSION THREAD

**Discussion by samvanemelen** [openglobus/openglobus#613] [General]
Hi all

I am playing around with openglobus and wanted to render a LOT of objects on the globus (custom trees, houses etc.). This is probably a going to be a battle for performance, so I was wondering what would be the most efficient way to do this, and what would be the pro's and cons of each option.

- Entities and 3Dobjects are probably out of the question because I believe rendering 3d objects is one of the most expensive operations. I also don't think we can add custom textures yet? 
- Then I thought about billboards. That should be quite performant but I am not sure if it is the look I want to go for.
- I also came across CanvasTiles which might be a great way to render a lot of things, because you do the canvas generation only once (on multiple threads), then it just shows the 'image'. This way you can bundle a lot of sprites in a single draw call. And this can also be animated with a LayerAnimation which is nice!
- And as a final thing I also saw you can use GeoTexture2d to render framebuffers directly. It sounds like it is something performant but I have no idea what this would entail.

I would love you here what others think of this. I know there isn't one right answer, but are there some pro's/cons that I missed with these options? did I get something wrong or are there other options I didn't think about?

**Comment by Zemledelec** [comment]
Hi @samvanemelen !

1) There is a GeoObject entity, it has an instancing interface, but we haven't implemented textures on it yet
2) CanvasTiles might be extended for 3d objects, but I've never thought about it. On the other hand, it loads quadtree nodes accordingly with the current camera attitude.
3) GeoTexture2d, the same as GeoVideo, and GeoImage (it has common parent) are visualized as images, It would help a lot if you want to render A LOT of 3d objects here.

I recommend taking a look at WebGL instancing API, and GeoObject entities. As far as opoenglobus has a direct way to WebGL API, it's easy to create your own scene, with your own shaders. 

P.S.
FYI @pavletto is working on geoobject texturing.

**Reply by pavletto** [reply]
yep,it will be ready soon. I would like to add that the choice of tool depends on what objects you want to display. they are all different or not (it depends on whether the instancing is suitable or not) how complex and detailed they are?

**Reply by program-sam** [reply]
Thanks for the insights! This is the kind of thing I am aiming for. I am trying to populate quite a lot. CanvasTiles seem to be holding up reasonably well for this. Render once on a worker thread and everything onwards is smooth because it is baked in a tile.  And by adding multiple CanvasTiles layers with different heights I can still create that parallax / 3d feel.

For this scale of objects I feel 3d objects would get either to slow, or too technical to do perfectly optimized that probably won't enjoy the process (or is it not as bad as I think?). Plus, the quadtree that tiles uses makes it really easy to only render what is relevant.

https://user-images.githubusercontent.com/75586344/209860391-71e83588-af5a-4d6c-b5ac-628d68bd0f15.mp4

**Comment by Zemledelec** [comment]
I suppose for instanced objects TileOffset in this way would be a very good solution! https://github.com/openglobus/openglobus/discussions/634

**Reply by program-sam** [reply]
Exactly, it works now as well, things are going nicely. But I can imagine performance being even better and overhead being lower If I could generate just have a handful of tiles at higher resolution instead of 20 smaller tiles. Especially at high zoom levels.

https://user-images.githubusercontent.com/75586344/223372704-52e09846-1a32-45db-8105-2ccb13f6feed.mp4
